test/framework/clustervalidator.go

```diff
@@ -0,0 +1,178 @@
+}
+
+// Validate runs through the set registered validations and returns an error if any of them fail after a number of retries.
+func (c *ClusterValidator) Validate(ctx context.Context) error {
+	for _, v := range c.validations {
+		err := retrier.Retry(v.maxRetries, v.backoffPeriod, func() error {
+			return v.validation(ctx, c.ClusterOpts)
+		})
+		if err != nil {
+			return fmt.Errorf("validation faild %v", err)
+		}
+	}
+
+	return nil
+}
+
+// Reset clears the registered validations.
+func (c *ClusterValidator) Reset() {
+	c.validations = []retriableValidation{}
+}
+
+// ClusterValidatorOpts represent the data to be passed as an argument to the Validate method.
+type ClusterValidatorOpts = func(cv *ClusterValidator)
+
+// NewClusterValidator returns a cluster validator.
+func NewClusterValidator(opts ...ClusterValidatorOpts) *ClusterValidator {
+	cv := ClusterValidator{
+		ClusterOpts: clusterOpts{},
+		validations: []retriableValidation{},
+	}
+
+	for _, opt := range opts {
+		opt(&cv)
+	}
+	return &cv
+}
+
+type clusterOpts struct {
+	ClusterClient          client.Client // the client for the cluster
+	ManagmentClusterClient client.Client // the client for the management cluster
+	ClusterSpec            *cluster.Spec // the cluster spec
+}
+
+func validateEKSAObjects(ctx context.Context, validateOpts clusterOpts) error {
+	for _, obj := range validateOpts.ClusterSpec.ChildObjects() {
+		u := &unstructured.Unstructured{}
+		u.SetGroupVersionKind(obj.GetObjectKind().GroupVersionKind())
+		key := types.NamespacedName{Namespace: obj.GetNamespace(), Name: obj.GetName()}
+
+		if key.Namespace == "" {
+			key.Namespace = "default"
+		}
+
+		err := validateOpts.ManagmentClusterClient.Get(ctx, key, u)
+		if err != nil {
+			return fmt.Errorf("cluster object does not exist %s", err)
+		}
+	}
+
+	return nil
+}
+
+func validateClusterReady(ctx context.Context, validateOpts clusterOpts) error {
+	capiCluster, err := controller.GetCAPICluster(ctx, validateOpts.ManagmentClusterClient, validateOpts.ClusterSpec.Cluster)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve cluster %s", err)
+	}
+
+	if capiCluster == nil && err == nil {
+		return fmt.Errorf("cluster %s does not exist", capiCluster.Name)
+	}
+
+	if err := checkClusterReady(capiCluster); err != nil {
+		return err
+	}
+	return nil
+}
+
+func validateControlPlaneNodes(ctx context.Context, validateOpts clusterOpts) error {
+	cpNodes := &corev1.NodeList{}
+	err := validateOpts.ClusterClient.List(ctx, cpNodes, client.MatchingLabels{"node-role.kubernetes.io/control-plane": ""})
+	if err != nil {
+		return fmt.Errorf("failed to list controlplane nodes %s", err)
+	}
+
+	cpConfig := validateOpts.ClusterSpec.Cluster.Spec.ControlPlaneConfiguration
+	if len(cpNodes.Items) != cpConfig.Count {
+		return fmt.Errorf("control plane node count does not match expected: %v of %v", len(cpNodes.Items), cpConfig.Count)
+	}
+
+	for _, node := range cpNodes.Items {
+		err := validateNodeReady(node)
```

Comment:
We are gonna need to add something else here
this validation only checks there is a determined number of nodes and all nodes are ready. But it doesn't check if those nodes are the nodes we expect. For example, we might be expecting 3 nodes, but after an upgrade, this validation will succeed with two old and one new node.

And easy way would be to check the kubernetes version. There are other things we should check, but I think that's the bare minimum to ensure nodes have been rolled out.